pkg/ipcache/metadata.go

```diff
@@ -327,43 +417,64 @@ func RemoveLabelsFromIPs(
 
 		triggerer.TriggerPolicyUpdates(false, "kube-apiserver identity updated by removal")
 	}
+	for ip, pair := range toReplace {
+		hIP, key := IPIdentityCache.getHostIPCache(ip)
+		meta := IPIdentityCache.getK8sMetadata(ip)
+		IPIdentityCache.deleteLocked(ip, pair.old.Source) // remove old entry and replace
```

Comment:
Why Delete? Why not just upsert?

This would introduce an additional time period in the datapath between executing this syscall and successfully removing the previous ipcache mapping for the IP, up until this process is next scheduled and completes through to the `upsertLocked()` call below. The identity used for such traffic in the middle period represents neither the old or new state, but instead would be a third identity (most likely something like world).

```suggestion
```

Reply:
(If the answer here is the source, maybe we can figure out a variant on upsertLocked() that can account for this, because we shouldn't be issuing two separate syscalls to make this transition happen due to the datapath impact)

Reply:
Yes this is because of the `source`. We cannot replace the `KubeAPIServer` source with an upsert if what we are upserting as a weaker source. This happen in the case when

1. an IP is associated with the kube-apiserver (entry in ipcache with src=kube-apiserver)
2. and a ToCIDR policy overlaps with the same IP
3. then IP from (1) is no longer associated with kube-apiserver, so ipcache entry from (2) must "take over"

Reply:
Synced offline. The approach is to have some way to indicate that we want to bypass the source check. In the future, we won't need something like this once we have https://github.com/cilium/cilium/issues/18301.